src/Geometry/MeshReader.h

```diff
@@ -375,6 +412,69 @@ class MeshReader
 		}
 	}
 
+	/*
+         * Exchanges halo tetrahedrons between MPI neighbors
+         * */
+	void exchangeGroupsWithMPINeighbors() {
+		// TODO: (if there is time) generalize halo exchange to arbitrary data maybe?
+		size_t numMPIDomains = m_MPINeighbors.size();
+
+		std::unordered_map<int, std::vector<int>> sendData;
+		std::vector<MPI_Request> sendRequests(numMPIDomains);
+		std::vector<MPI_Request> recvRequests(numMPIDomains);
+
+		constexpr int exchangeTag{11};
+		auto communicator = seissol::MPI::mpi.comm();
+
+		auto &domainElements = m_elements;
+
+		std::unordered_map<int, std::vector<int>> recvData;
+
+		size_t counter{};
+		for (auto it = m_MPINeighbors.begin(); it != m_MPINeighbors.end(); ++it, ++counter) {
+
+			auto neighborRank = it->first;
+			auto numElements = it->second.elements.size();
+			const size_t messageSize = numElements;
+
+			sendData[neighborRank].resize(messageSize);
+			recvData[neighborRank].resize(messageSize);
+
+			auto* startAddress = &(recvData[neighborRank][0]);
+			MPI_Irecv(const_cast<int*>(startAddress),
+								messageSize,
+								MPI_INT,
+								neighborRank,
+								exchangeTag,
+								communicator,
+								&recvRequests[counter]);
+
+			for (size_t elementIdx = 0; elementIdx < numElements; ++elementIdx) {
+				auto localElementIdx = it->second.elements[elementIdx].localElement;
+				auto& element = domainElements[localElementIdx];
+
+				sendData[neighborRank][elementIdx] = element.group;
+			}
+
+			MPI_Isend(const_cast<int*>(&sendData[neighborRank][0]),
+								messageSize,
+								MPI_INT,
+								neighborRank,
+								exchangeTag,
+								communicator,
+								&sendRequests[counter]);
+		}
+
+		std::vector<MPI_Status> sendStatus{numMPIDomains};
+		std::vector<MPI_Status> recvStatus{numMPIDomains};
+
+		MPI_Waitall(numMPIDomains, &sendRequests[0], &sendStatus[0]);
+		MPI_Waitall(numMPIDomains, &recvRequests[0], &recvStatus[0]);
```

Comment:
For the group transfer, it's basically the code from the method above it for the vertex transfer (the two methods could be merged, maybe); i.e. just copied and adjusted legacy code. :) But ok, it can be re-written to be a bit prettier.